doc/1.4/language.md

```diff
@@ -3085,48 +3191,270 @@ return m(e1)
+the method must be of [*serializable type*](#serializable-types), unless that
+input parameter receives a message component.
+This means that after-on-hook statements cannot be used with methods that e.g.
+have pointer input parameters, unless the arguments for those input parameters
+are message component parameters of the `after`.
 
-**Note:** We plan to extend the `after` statement to allow for users to
-explicitly state what objects the posted events are to be associated with.
+Example use:
+```
+hook (int, float) h;
 
-</div>
+method my_callback(int i, float f, bool b) {
+    ...
+}
+
+method m() {
+    after h -> (x, y): my_callback(x, y, false);
+}
+
+method send_message() {
+    // Assuming m() has been called once before, this 'send_now' will result in
+    // `my_callback(1, 3.7, false)` being called.
+    h.send_now(1, 3.7);
+}
+```
+
+If the hook has only one message component, the syntax <tt>-> <em>msg</em></tt>
+can be used instead, and if the hook has no message components, then the binding
+syntax can be entirely omitted. Any message component parameter can be used for
+any number of arguments, but cannot be used as anything *but* a direct argument.
+For example, using the definitions of `h` and `my_callback` as above, the
+following use of `after` is valid:
+```
+after h -> (x, y): my_callback(x, x, false)
+```
+Note that the first message component is used multiple times, and the second
+is not used at all.
+
+In contrast, the following use of `after` is invalid:
+```
+after h -> (x, y): my_callback(i, y + 1.5, false)
+```
+as the message component parameter `y` is used, but not as a direct argument.
+
+#### Immediate After Statements
+<pre>
+after: <em>method</em>(<em>e1</em>, ... <em>eN</em>);
+</pre>
+
+In this form, the specified point in the future is when all entries into the
+device currently on the call stack have been completed, and control would
+otherwise return to the simulation engine (or whatever caused the outermost
+entry into the device). Semantically, this makes immediate after statements very
+similar to `after 0 s: ...` but has a number of advantages. Most notably,
+immediate after statements does not incur the same performance penalty that
+repeated use of `after 0 s: ...` does, as immediate after does not make use of
+Simics events.
+
+Immediate after statements are most useful to avoid ordering bugs. It can be
+used to delay a method call until all current lines of execution into the device
+have been completed, and the device is guaranteed to be in a consistent state
+where it is ready to handle the method call.
+
+For example, consider the following device, representing an i2c master connected
+to a slave via `i2c_link`, and receiving responses from the slave via the
+`i2c_in` port. Its function is simple: once prompted, the device will start the
+slave, and then continually read packets from it until done, handling the
+asynchronous logic through an FSM.
+```
+connect i2c_link {
```

Comment:
This example is _too long_! It also assumes that the user has knowledge about the `i2c_master_v2` and `i2c_slave_v2` interfaces. Issue is that we need to present an example of an ordering bug realistic enough that _people might recognize it_. So...

1. We need an interaction with something external. That means connect+implement
2. The device needs to have some associated state, where in one state it can handle a particular interface call, but in another it can't.
3. The state logic that prompts the external device should be separate from the implement handling its own interface calls, and it's within the former logic that the ordering bug happens...
4. ... And the ordering bug can be resolved by using immediate after _in the implement_, without having to touch the state logic.

These are _pretty_ hard to satisfy simultaneously while keeping it short. We'll have to think about this.

Reply:
I still am not fond of this example, and I don't know how to best address it. For the moment I've mitigated this issue by presenting it as the final part of this subsubsubsection, and describing it as a scenario rather than a simple example.

Reply:
Perhaps some wire fiddling over the signal interface? Raise an outgoing signal A toward a power management device, and wait for incoming signal B to be raised as an ack that it's OK to proceed with some operation (and lower A again). Similar patterns are pretty common in real HW.

Reply:
Example redone, and you gave it a thumbs up, so I'm keeping it (modulo the use of a `port`).